tap/source/packet_source_manager.go

```diff
@@ -0,0 +1,140 @@
+
+func NewPacketSourceManager(procfs string, pids string, filename string, interfaceName string,
+	behaviour TcpPacketSourceBehaviour) (*PacketSourceManager, error) {
+	sources := make([]*tcpPacketSource, 0)
+	host_source, err := newHostPacketSource(filename, interfaceName, behaviour)
+
+	if err != nil {
+		return nil, err
+	}
+
+	sources = append(sources, host_source)
+
+	if pids != "" {
+		netnsSources := newNetnsPacketSources(procfs, pids, interfaceName, behaviour)
+		sources = append(sources, netnsSources...)
+	}
+
+	return &PacketSourceManager{
+		sources: sources,
+	}, nil
+}
+
+func newHostPacketSource(filename string, interfaceName string,
+	behaviour TcpPacketSourceBehaviour) (*tcpPacketSource, error) {
+	var name string
+
+	if filename == "" {
+		name = fmt.Sprintf("host-%v", interfaceName)
+	} else {
+		name = fmt.Sprintf("file-%v", filename)
+	}
+
+	source, err := newTcpPacketSource(name, filename, interfaceName, behaviour)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return source, nil
+}
+
+func newNetnsPacketSources(procfs string, pids string, interfaceName string,
+	behaviour TcpPacketSourceBehaviour) []*tcpPacketSource {
+	result := make([]*tcpPacketSource, 0)
+
+	for _, pidstr := range strings.Split(pids, ",") {
+		pid, err := strconv.Atoi(pidstr)
+
+		if err != nil {
+			logger.Log.Errorf("Invalid PID: %v - %v", pid, err)
+			continue
+		}
+
+		nsh, err := netns.GetFromPath(fmt.Sprintf("%v/%v/ns/net", procfs, pid))
+
+		if err != nil {
+			logger.Log.Errorf("Unable to get netns of pid %v - %v", pid, err)
+			continue
+		}
+
+		src, err := newNetnsPacketSource(pid, nsh, interfaceName, behaviour)
+
+		if err != nil {
+			logger.Log.Errorf("Error starting netns packet source for %v - %v", pid, err)
+			continue
+		}
+
+		result = append(result, src)
+	}
+
+	return result
+}
+
+func newNetnsPacketSource(pid int, nsh netns.NsHandle, interfaceName string,
+	behaviour TcpPacketSourceBehaviour) (*tcpPacketSource, error) {
+
+	done := make(chan *tcpPacketSource)
+	errors := make(chan error)
+
+	go func(done chan<- *tcpPacketSource) {
+		// Setting a netns should be done from a dedicated OS thread.
+		//
+		// goroutines are not really OS threads, we try to mimic the issue by
+		//	locking the OS thread to this goroutine
+		//
+		runtime.LockOSThread()
+		defer runtime.UnlockOSThread()
+
+		if err := netns.Set(nsh); err != nil {
+			logger.Log.Errorf("Unable to set netns of pid %v - %v", pid, err)
+			errors <- err
+			return
+		}
+
+		name := fmt.Sprintf("netns-%v-%v", pid, interfaceName)
+		src, err := newTcpPacketSource(name, "", interfaceName, behaviour)
+
+		if err != nil {
+			logger.Log.Errorf("Error listening to PID %v - %v", pid, err)
+			errors <- err
+			return
+		}
+
+		done <- src
+	}(done)
+
+	select {
+	case err := <-errors:
+		return nil, err
+	case source := <-done:
+		return source, nil
+	}
+}
+
+func (m *PacketSourceManager) ReadPackets(ipdefrag bool, packets chan<- TcpPacketInfo) {
+	for _, src := range m.sources {
```

Comment:
From what I know the number of concurrent packets is implicitly limited by the assembler, via its memory limit. All sources share the same assembler so the limit apply to all of them together. Please let me know if I'm missing a place in the code that explicitly limit the goroutines.

`m.sources` should contain at least one source for the host, as initialized in the `NewPacketSourceManager` function.

Reply:
> From what I know the number of concurrent packets is implicitly limited by the assembler, via its memory limit. All sources share the same assembler so the limit apply to all of them together. Please let me know if I'm missing a place in the code that explicitly limit the goroutines.

There is no limit for the number of goroutines. It's just under a high traffic load, the tapper has a chance to getting `OOMKilled`. I'm not sure if the situation gets worse with this change. Without knowing the possible values of `pids`, it's hard to tell.

> `m.sources` should contain at least one source for the host, as initialized in the `NewPacketSourceManager` function.

Yes, I'm aware of that. `len(m.sources)` is `1` in case of `pids` equals to `""`.

Reply:
We shouldn't have more than few PIDs - i can't tell exactly how many as it depend on various things. In any case, the resources limits shared across all the resources

Reply:
It's less about the number of PIDs but more about what are we really listening into. Like are we going to capture some irrelevant traffic as well? (once the implementation of this feature really complete) Because that would cost CPU time and spike the memory usage.

`m.sources` being a slice rather than a single source, raises some questions that cannot be answered without the **integration being complete**. For example;

- What determines the length of `m.sources`? Number of pods or containers in a single node? If so, do we get duplicate data if the traffic is happening between two pods inside the same machine?
- Is there traffic in service mesh caused by liveness probes or something else? How frequent are they? Will it degrade the performance of the tapper.

Reply:
For now - m.sources is exactly 1, containing the host tapper, like the current behaviour, Unless someone pass the `pids` which no one should for real scenarios. 

In the next phase when we'll implement the Envoy auto discoverer, we'll talk about the performance and all impacts. We'll probably communicate it to the user somehow, but this is something that @haiut may elaborate.

Reply:
[Similarly](https://github.com/up9inc/mizu/pull/418#discussion_r742736006). Let's see the answers in the following PRs.